libs/partners/astradb/langchain_astradb/vectorstores/astradb.py

```diff
@@ -0,0 +1,1339 @@
+      it is suggested to keep the quantity
+          bulk_insert_batch_concurrency * bulk_insert_overwrite_concurrency
+      much below 1000 to avoid exhausting the client multithreading/networking
+      resources. The hardcoded defaults are somewhat conservative to meet
+      most machines' specs, but a sensible choice to test may be:
+          bulk_insert_batch_concurrency = 80
+          bulk_insert_overwrite_concurrency = 10
+      A bit of experimentation is required to nail the best results here,
+      depending on both the machine/network specs and the expected workload
+      (specifically, how often a write is an update of an existing id).
+      Remember you can pass concurrency settings to individual calls to
+      add_texts and add_documents as well.
+
+      A note on passing astra_db_client and/or async_astra_db_client instead
+      of the credentials (token, api_endpoint):
+      - if you pass only the async client when creating the store,
+        the sync methods will error when called.
+      - conversely, if you pass only the sync client, the async methods will
+        still be available, but will be wrapping its sync counterpart
+        in a `run_in_executor` construct instead of using the native async.
+    """
+
+    @staticmethod
+    def _filter_to_metadata(filter_dict: Optional[Dict[str, Any]]) -> Dict[str, Any]:
+        if filter_dict is None:
+            return {}
+        else:
+            metadata_filter = {}
+            for k, v in filter_dict.items():
+                if k and k[0] == "$":
+                    if isinstance(v, list):
+                        metadata_filter[k] = [
+                            AstraDBVectorStore._filter_to_metadata(f) for f in v
+                        ]
+                    else:
+                        # assume each list item can be fed back to this function
+                        metadata_filter[k] = AstraDBVectorStore._filter_to_metadata(v)  # type: ignore[assignment]
+                else:
+                    metadata_filter[f"metadata.{k}"] = v
+
+            return metadata_filter
+
+    def __init__(
+        self,
+        *,
+        embedding: Embeddings,
+        collection_name: str,
+        token: Optional[str] = None,
+        api_endpoint: Optional[str] = None,
+        astra_db_client: Optional[AstraDBClient] = None,
+        async_astra_db_client: Optional[AsyncAstraDBClient] = None,
+        namespace: Optional[str] = None,
+        metric: Optional[str] = None,
+        batch_size: Optional[int] = None,
+        bulk_insert_batch_concurrency: Optional[int] = None,
+        bulk_insert_overwrite_concurrency: Optional[int] = None,
+        bulk_delete_concurrency: Optional[int] = None,
+        pre_delete_collection: bool = False,
+    ) -> None:
+        """
+        Create an AstraDBVectorStore vector store object. See class docstring for help.
+        """
+        try:
+            from astrapy.db import (
```

Comment:
all imports can now be top-level as a concrete dep! No more optional needed, and can properly lint too :)